### PR TITLE
Switches from pipenv and requirements.txt to poetry to manage dependencies

### PR DESCRIPTION
**Pull request by SaptakS**
Refs #1094 

This needs cross-platform testing. I have tested it in linux and seems to be working great. I am on my way to test with windows. @micahflee it would be great if you can test in mac OSx.

I have also for the time kept a fixed version in the dependency instead of using `^` in the version because that's how it was in requirements.txt and I didn't want to bring any breaking changes by allowing updates. But if it is safe to do `^` then I will add that in `pyproject.toml`

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,50 @@
+[tool.poetry]
+name = "onionshare"
+version = "2.2"
```
Comment (micahflee) [inline pyproject.toml]: Can you add a note to the release documentation (https://github.com/micahflee/onionshare/blob/develop/BUILD.md#changelog-version-and-signed-git-tag) to update the version in this file with each release?

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+description = "OnionShare lets you securely and anonymously send and receive files. It works by starting a web server, making it accessible as a Tor onion service, and generating an unguessable web address so others can download files from you, or upload files to you. It does _not_ require setting up a separate server or using a third party file-sharing service."
+authors = ["Micah Lee <micah.lee@theintercept.com>"]
+license = "GPLv3+"
+
+[tool.poetry.dependencies]
+python = "^3.7"
```
Comment (micahflee) [inline pyproject.toml]: Actually sadly there's a bug where onionshare doesn't work with python 3.8 (#1079) though I'd love to fix that. So this should just be "3.7"

Comment (micahflee) [reply pyproject.toml]: Or, however you do >= 3.7 and < 3.8

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+atomicwrites = "1.3.0"
+attrs = "19.1.0"
+more-itertools = "7.2.0"
+pluggy = "0.13.0"
+py = "1.8.0"
+pytest = "5.1.2"
+pytest-faulthandler = "2.0.1"
+pytest-qt = "3.2.2"
+six = "1.12.0"
+urllib3 = "1.25.3"
```
Comment (micahflee) [inline pyproject.toml]: None of these are tied to any specific version, they can all have `^` -- and while you're at it, I would be fine bumping all of the versions to their newest versions, so long as the tests pass

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,50 @@
+[tool.poetry]
+name = "onionshare"
+version = "2.2"
+description = "OnionShare lets you securely and anonymously send and receive files. It works by starting a web server, making it accessible as a Tor onion service, and generating an unguessable web address so others can download files from you, or upload files to you. It does _not_ require setting up a separate server or using a third party file-sharing service."
+authors = ["Micah Lee <micah.lee@theintercept.com>"]
```
Comment (micahflee) [inline pyproject.toml]: Can you change my email here to micah@micahflee.com? :)

File: pyproject.toml (hunk 3 above)
Comment (micahflee) [reply pyproject.toml]: You can update to all the latest versions with `poetry update`

File: pyproject.toml (hunk 2 above)
Comment (micahflee) [reply pyproject.toml]: Actually, I'm hoping that upgrading to stem 1.8.0 will allow onionshare will work with python 3.8 as well. If that's the case, then this PR will also fix #1079

Comment (SaptakS) [reply pyproject.toml]: Can do that in this PR itself.

File: pyproject.toml (hunk 3 above)
Comment (micahflee) [reply pyproject.toml]: Actually, while troubleshooting in #1079 I discovered that if you set the version for all of the dependencies to `*`, and then run `poetry install` and `poetry update`, it updates all dependencies in `poetry.lock`, and fixes the python 3.8/stem problem.

File: pyproject.toml (hunk 1 above)
Comment (SaptakS) [reply pyproject.toml]: Done

File: pyproject.toml (hunk 3 above)
Comment (SaptakS) [reply pyproject.toml]: Changed everything to `*` and then did a `poetry update` to update the `poetry.lock`

**Comment by SaptakS** [comment]
@micahflee I haven't done the linux build part yet. Do you mean we should use poetry instead of apt and dnf for python dependencies in `BUILD.md`? I am not super clear on what you mean. :sweat_smile:

**Pull request merged**